lib/Backend/JITThunkEmitter.cpp

```diff
@@ -0,0 +1,285 @@
+        if (localAddress == nullptr)
+        {
+            return;
+        }
+        UnprotectPage(localAddress);
+        memset(localAddress, 0xCC, ThunkSize);
+        ProtectPage(localAddress);
+        this->codeAllocator->FreeLocal(localAddress);
+    }
+    FlushInstructionCache(this->processHandle, (PVOID)thunkAddress, ThunkSize);
+}
+
+template <typename TAlloc> inline
+uintptr_t
+JITThunkEmitter<TAlloc>::EnsureInitialized()
+{
+    if (this->baseAddress != NULL)
+    {
+        return this->baseAddress;
+    }
+
+    // only take a lock if we need to initialize
+    {
+        AutoCriticalSection autoCs(&this->cs);
+        // check again because we did the first one outside of lock
+        if (this->baseAddress == NULL)
+        {
+            this->baseAddress = (uintptr_t)this->codeAllocator->Alloc(nullptr, TotalThunkSize, MEM_RESERVE, PAGE_EXECUTE, true);
+        }
+    }
+    return this->baseAddress;
+}
+
+template <typename TAlloc> inline
+bool
+JITThunkEmitter<TAlloc>::IsInThunk(uintptr_t address) const
+{
+    return IsInThunk(this->baseAddress, address);
+}
+
+/* static */
+template <typename TAlloc> inline
+bool
+JITThunkEmitter<TAlloc>::IsInThunk(uintptr_t thunkBaseAddress, uintptr_t address)
+{
+    bool isInThunk = address >= thunkBaseAddress && address < thunkBaseAddress + TotalThunkSize;
+    Assert(!isInThunk || address % ThunkSize == 0);
+    return isInThunk;
+}
+
+/* static */
+template <typename TAlloc> inline
+void
+JITThunkEmitter<TAlloc>::EncodeJmp(char * localPageAddress, uintptr_t thunkAddress, uintptr_t targetAddress)
+{
+    char * localAddress = localPageAddress + thunkAddress % AutoSystemInfo::PageSize;
+    ptrdiff_t relativeAddress = targetAddress - thunkAddress - DirectJmpIPAdjustment;
+#if _M_X64
+    if (relativeAddress > INT_MAX || relativeAddress < INT_MIN)
+    {
+        memcpy_s(localAddress, ThunkSize, IndirectJmp, ThunkSize);
+        uintptr_t * jmpTarget = (uintptr_t*)(localAddress + IndirectJmpTargetOffset);
+        *jmpTarget = targetAddress;
+    }
+    else
+#endif
+    {
+        memcpy_s(localAddress, ThunkSize, DirectJmp, ThunkSize);
+        uintptr_t * jmpTarget = (uintptr_t*)(localAddress + DirectJmpTargetOffset);
+        *jmpTarget = relativeAddress;
+    }
+}
+
+template <typename TAlloc> inline
+bool
+JITThunkEmitter<TAlloc>::IsThunkPageEmpty(uintptr_t address) const
+{
+    Assert(address == GetThunkPageStart(address));
+    BVIndex pageStartIndex = GetThunkIndexFromAddress(address);
+    Assert(pageStartIndex != BVInvalidIndex);
+    BVStatic<ThunksPerPage> * pageBV = this->freeThunks.GetRange<ThunksPerPage>(pageStartIndex);
+    return pageBV->IsAllSet();
+}
+
+template <> inline
+void
+JITThunkEmitter<VirtualAllocWrapper>::ProtectPage(void * address)
+{
+    DWORD oldProtect;
+    VirtualProtectEx(this->processHandle, address, ThunkSize, PAGE_EXECUTE, &oldProtect);
+    AssertOrFailFast(oldProtect == PAGE_EXECUTE_READWRITE);
```

Comment:
>oldProtect == PAGE_EXECUTE_READWRITE [](start = 21, length = 36)

check return value as well, or initialize oldProtect, in case VirtualProtectEx failed